File: Python/290-Word-Pattern.py
class Solution:
    '''
    https://leetcode.com/problems/word-pattern/
    '''

    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        map = {}
        if len(words) != len(pattern) or len(set(pattern)) != len(set(words)):
            return False

        for idx, ch in enumerate(pattern):
            if ch in map:
                if map[ch] != words[idx]:
                    return False
            else:
                map[ch] = words[idx]
        return True


s = Solution()
print(s.wordPattern("abba", "dog cat cat dog"))
print(s.wordPattern("abba", "dog cat cat fish"))
print(s.wordPattern("abba", "dog dog dog dog"))
print(s.wordPattern("abc", "b c a"))

File: JavaScript/290-Word-Pattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function (pattern, s) {
    var words = s.split(" ")
    var map = new Map()
    if ((words.length !== pattern.length)) return false
    if (new Set(words).size !== new Set(pattern).size) return false

    for (var i = 0; i < words.length; i++) {
        var ch = pattern.charAt(i)
        var w = words[i]
        if (ch in map) {
            if (map[ch] != w) return false
        } else
            map[ch] = w;
    }
    return true
};

console.log(wordPattern("abba", "dog cat cat dog"))